### DISCUSSION TITLE
Project direction and dropping globals

### DISCUSSION THREAD

**Discussion by filasieno** [sunbains/embedded-innodb#62] [General]
## Project Direction 

I took a some time to look at the code and the last commits.
I'd like a embedded DB for a project of mine and starting from innodb seems very interesting for me.
I'd like to understand If the direction you are going is compatible with my needs or if I could adapt my needs to share the project.
I need to support a code transformation system, with persistent storage. Massive size.

If the direction is embedded, my personal ideal would be totally stateless.
- No malloc (literally, in the sense not even malloc/realloc/free fn pointers replacement); just pass in a memory block.
- No globals (and static variables) with minor exceptions (maybe counters, etc ...)

Embedding becomes easy this way.
I prefer a [Game Programming style C++](https://www.youtube.com/watch?v=rX0ItVEVjHc) with few templates, no virtuals, and data orientation with an official [command style public C API](https://github.com/unclebob/cmuratori-discussion/blob/main/cleancodeqa-2.md). 

Could you tell me a bit of the project direction ? Simply I do not want to waste your time.

## Globals

I have a proposal for globals and state management which I think is superior as design and makes it way easier to manage the complex interactions that a DB requires; for instance to avoid things like this:

```C++
srv_buf_pool->m_flusher->batch(srv_dblwr, BUF_FLUSH_LIST, PCT_IO(100), IB_UINT64_T_MAX);
```

3 components: DoubleWriter, BufferPool, Flusher, ... then other components such as FSP which again uses the buffer pool; I think that the component structure is essential to manage complexity but encapsulation is harming in this case.

A simpler AND more powerful solution would be something like.
Just to explain the concept (types are invented)

```C++
struct BufferPoolState { 
      // BufferPool state variables 
}

struct LogWriterState {
     // LogWriter state variables
}

struct TxManagerState {
     // TxManager state variables
}

struct AnyOtherManagerState { 
    // AnyOtherManager state variables
};

struct InnoDBState {
     BufferPoolState buffPoolMngr;
     LogWriterState lgwrMngr;
     TxManagerState txMngr;
     AnyOtherManagerState aoMngr;
};
```

The State part of the components would have methods that:
 - depend only on self 
 - never receive other components as arguments
basically each Component State offers primitives for local state mutation.

And then a set of __component interfaces__:

```C++
class BufferPool {
     // Buffer Pool routines
     InnoDBState *m_state;
};

class LogWriter {
     // LogWriter routines
     InnoDBState *m_state;
};

class TxManager {
     // TxManager manager
     InnoDBState *m_state;
};

class AnyOtherManagerState { 
    // 
   InnoDBState *m_state;
};

class InnoDB { 
    // 
   InnoDBState *m_state;
};
```

Basically you bind the __component interface__ to the state.

A component can touch anything if required, but you can isolate the different component logic making the DB manageable.
Each component can be approached as a unit AND It's easier for a component to manage other related component state if required.

Clear separation of data and routines makes complex transformations easier.

**Comment by sunbains** [comment]
What you see in the code is an interim state. It's not the final state. What you are proposing is something we can consider once the globals are identified. They are sprinkled all over the code and once everything is a class and we have the dependencies tied sown,

e.g., Buffer_pool->Flushing->Fil->IO, or Btree->{FSP, Lock_sys} etc.

Then we can slice and dice as we please.

**Reply by filasieno** [reply]
Ok.

Just to be clear, as global you also mean static variables or just the syntactical form of a "C global"? 

It seems to me that your first objective is just to package the C routines into C++ classes.

Few globals will remain which probably are going to be the different top managers. All the rest packaged in classes.

Correct?

**Reply by sunbains** [reply]
There are too many global variables to list here, status variables, config variables, handles to different components and their internal state etc.  Syntactically C and C++ global variables are the same.

 So, yes, convert all the different components into C++ structs (I don't like class/private/protected etc.  unless absolutely necessary). After this is done, there shouldn't be any global variable. Everything should be reachable via an InnoDB instance only, which is passed to different components during construction. Then the fun starts.

**Reply by filasieno** [reply]
A static variable is a global variable form my point of view; in this sense I was asking.

I dont't think that by removing globals you mean moving them to static variables in classes; but in general to make the whole system reentrant. 

For instance: **it should be possible to have more than one full database initialized on the stack in the user's main function. To do this effectively we need to remove both globals and static variables; very similar to what happens with parsers/parser generators.**

correct?

If this is established, than I have one last question and I can begin to send PRs.

Sorry if I'm being pedantic; I just mean to clearly establish your objectives to move in the direction you wish to go.
Also, I think that this initial step you are pursuing has great value.

**Reply by sunbains** [reply]
Yes, statics are global too.remove them too.

**Reply by filasieno** [reply]
Agreed. 

I'll send PRs and ask questions later.

Let's see how it goes.

**Comment by sunbains** [comment]
There is no project direction as such at the moment, except to clean it up, port all the major optimizations from 5.6 and 5.7. I want to perform at par (or better) than 5.7 and then decide which limitations to tackle. 5.6 and 5.7 have very critical  optimization for improving concurrency and scalability, since I wrote them they shouldn't be too hard.

**Comment by sunbains** [comment]
I don't want to change the API because the API has been tested in MySQL itself for the memcache plugin. I will make it more C++ from its current C style. This code was originally all C.
